From 66b5e07a3cd5c7e14fe97388000e96ca41efc3f9 Mon Sep 17 00:00:00 2001
From: David Tarditi <dtarditi@microsoft.com>
Date: Tue, 31 Jul 2018 17:20:17 -0700
Subject: [PATCH] Establish ARM64 baseline. (#30)

* Establish testing/build baseline for samples on ARM64.

* Add mising detour_find_jmp_bounds for ARM64.
---
 samples/Makefile | 21 +++++++++++++++++++--
 src/detours.cpp  | 13 +++++++++++++
 2 files changed, 32 insertions(+), 2 deletions(-)

diff --git a/samples/Makefile b/samples/Makefile
index 6aee1ef..adaa8fe 100644
--- a/samples/Makefile
+++ b/samples/Makefile
@@ -39,8 +39,10 @@ all:
     @$(MAKE) /NOLOGO /$(MAKEFLAGS)
     cd "$(MAKEDIR)\echo"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS)
+!IF "$(DETOURS_TARGET_PROCESSOR)" != "ARM64"
     cd "$(MAKEDIR)\einst"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS)
+!ENDIF
 !IF "$(DETOURS_TARGET_PROCESSOR)" == "X86"
     cd "$(MAKEDIR)\excep"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS)
@@ -49,9 +51,11 @@ all:
     @$(MAKE) /NOLOGO /$(MAKEFLAGS)
     cd "$(MAKEDIR)\commem"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS)
+!IF "$(DETOURS_TARGET_PROCESSOR)" != "ARM64"
     cd "$(MAKEDIR)\findfunc"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS)
-!IF "$(DETOURS_TARGET_PROCESSOR)" != "ARM"
+!ENDIF
+!IF "$(DETOURS_TARGET_PROCESSOR)" != "ARM" && "$(DETOURS_TARGET_PROCESSOR)" != "ARM64"
     cd "$(MAKEDIR)\member"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS)
 !ENDIF
@@ -77,7 +81,7 @@ all:
     @$(MAKE) /NOLOGO /$(MAKEFLAGS)
     cd "$(MAKEDIR)\tracelnk"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS)
-!IF "$(DETOURS_TARGET_PROCESSOR)" != "ARM"
+!IF "$(DETOURS_TARGET_PROCESSOR)" != "ARM" && "$(DETOURS_TARGET_PROCESSOR)" != "ARM64"
     cd "$(MAKEDIR)\tryman"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS)
 !ENDIF
@@ -218,34 +222,42 @@ test:
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
     cd "$(MAKEDIR)\simple"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
+!IF "$(DETOURS_TARGET_PROCESSOR)" != "ARM64"
     cd "$(MAKEDIR)\slept"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
     cd "$(MAKEDIR)\setdll"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
     cd "$(MAKEDIR)\withdll"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
+!ENDIF
 !IF "$(DETOURS_TARGET_PROCESSOR)" == "X86"
     cd "$(MAKEDIR)\cping"
 #   @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
 !ENDIF
     cd "$(MAKEDIR)\disas"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
+!IF "$(DETOURS_TARGET_PROCESSOR)" != "ARM64"
     cd "$(MAKEDIR)\dtest"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
+!ENDIF
     cd "$(MAKEDIR)\dumpe"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
     cd "$(MAKEDIR)\dumpi"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
     cd "$(MAKEDIR)\echo"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
+!IF "$(DETOURS_TARGET_PROCESSOR)" != "ARM64"
     cd "$(MAKEDIR)\einst"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
+!ENDIF
 !IF "$(DETOURS_TARGET_PROCESSOR)" == "X86"
     cd "$(MAKEDIR)\excep"
 #   @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
 !ENDIF
+!IF "$(DETOURS_TARGET_PROCESSOR)" != "ARM64"
     cd "$(MAKEDIR)\comeasy"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
+
     cd "$(MAKEDIR)\commem"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
     cd "$(MAKEDIR)\findfunc"
@@ -254,10 +266,12 @@ test:
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
     cd "$(MAKEDIR)\region"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
+!ENDIF
 !IF "$(DETOURS_TARGET_PROCESSOR)" == "X64" || "$(DETOURS_TARGET_PROCESSOR)" == "IA64"
     cd "$(MAKEDIR)\talloc"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
 !ENDIF
+!IF "$(DETOURS_TARGET_PROCESSOR)" != "ARM64"
     cd "$(MAKEDIR)\traceapi"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
     cd "$(MAKEDIR)\tracebld"
@@ -268,12 +282,15 @@ test:
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
     cd "$(MAKEDIR)\traceser"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
+!ENDIF
 #    cd "$(MAKEDIR)\tracessl"
 #    @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
 #    cd "$(MAKEDIR)\tracetcp"
 #    @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
+!IF "$(DETOURS_TARGET_PROCESSOR)" != "ARM64"
     cd "$(MAKEDIR)\tracelnk"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
+!ENDIF
     cd "$(MAKEDIR)\impmunge"
     @$(MAKE) /NOLOGO /$(MAKEFLAGS) test
     cd "$(MAKEDIR)"
diff --git a/src/detours.cpp b/src/detours.cpp
index 8345c4d..3cd0e9d 100644
--- a/src/detours.cpp
+++ b/src/detours.cpp
@@ -974,6 +974,19 @@ inline PBYTE detour_skip_jmp(PBYTE pbCode, PVOID *ppGlobals)
     return pbCode;
 }
 
+inline void detour_find_jmp_bounds(PBYTE pbCode,
+                                   PDETOUR_TRAMPOLINE *ppLower,
+                                   PDETOUR_TRAMPOLINE *ppUpper)
+{
+    // We have to place trampolines within +/- 2GB of code.
+    ULONG_PTR lo = detour_2gb_below((ULONG_PTR)pbCode);
+    ULONG_PTR hi = detour_2gb_above((ULONG_PTR)pbCode);
+    DETOUR_TRACE(("[%p..%p..%p]\n", lo, pbCode, hi));
+
+    *ppLower = (PDETOUR_TRAMPOLINE)lo;
+    *ppUpper = (PDETOUR_TRAMPOLINE)hi;
+}
+
 inline BOOL detour_does_code_end_function(PBYTE pbCode)
 {
     ULONG Opcode = fetch_opcode(pbCode);
-- 
2.46.0.windows.1

